62.CAD/vertex_shader.hlsl

```diff
@@ -418,6 +407,77 @@ PSInput main(uint vertexID : SV_VertexID)
+        {
+            bool isInDrawSection = true;
+            if(lineStyle.hasStipples())
+            {
+                StyleAccessor styleAccessor;
+                styleAccessor.styleIdx = mainObj.styleIdx;
+
+                const float phaseShift = vk::RawBufferLoad<float>(drawObj.geometryAddress + sizeof(double2) + sizeof(float2) + sizeof(float), 8u);
+                const float normalizedPlaceInPattern = frac(phaseShift);
+                const uint32_t patternIdx = nbl::hlsl::upper_bound(styleAccessor, 0, lineStyle.stipplePatternSize, normalizedPlaceInPattern);
+                // odd patternIdx means a "no draw section" and current candidate should split into two nearest draw sections
+                isInDrawSection = !(patternIdx & 0x1);
+            }
+
+            if (isInDrawSection)
+            {
+                outV.setColor(float4(0.0f, 1.0f, 0.0f, 0.5f));
+                //outV.setColor(lineStyle.color);
+                const float lineThickness = screenSpaceLineWidth / 2.0f;
+                outV.setLineThickness(lineThickness);
+                
+                double2 circleCenter = vk::RawBufferLoad<double2>(drawObj.geometryAddress, 8u);
+                float2 v = vk::RawBufferLoad<float2>(drawObj.geometryAddress + sizeof(double2), 8u);
+                float cosAngleDifferenceHalf = vk::RawBufferLoad<float>(drawObj.geometryAddress + sizeof(double2) + sizeof(float2), 8u);
+
+                float2 vScreenSpace = float2(v.x, -v.y); // TODO: just do that in cpp (CPolyline::preprocessPolylineWithStyle)?
+                float2 circleCenterScreenSpace = transformPointScreenSpace(clipProjectionData.projectionToNDC, circleCenter);
+
+                if (vertexIdx == 0u)
+                {
+                    const float sinAngleDifferenceHalf = sqrt(1.0f - (cosAngleDifferenceHalf * cosAngleDifferenceHalf));
+                    const float32_t2x2 rotationMatrix = float32_t2x2(cosAngleDifferenceHalf, -sinAngleDifferenceHalf, sinAngleDifferenceHalf, cosAngleDifferenceHalf);
+                    const float2 v1ScreenSpace = normalize(mul(vScreenSpace, rotationMatrix)) * lineThickness;
+
+                    outV.position = float4(circleCenterScreenSpace + v1ScreenSpace, 0.0f, 1.0f);
+                }
+                else if (vertexIdx == 1u)
+                {
+                    outV.position = float4(circleCenterScreenSpace, 0.0f, 1.0f);
+                }
+                else if (vertexIdx == 2u)
+                {
+                    outV.position = float4(circleCenterScreenSpace + vScreenSpace * lineThickness, 0.0f, 1.0f);
+                }
+                else if (vertexIdx == 3u)
+                {
+                    const float sinAngleDifferenceHalf = sqrt(1.0f - (cosAngleDifferenceHalf * cosAngleDifferenceHalf));
+                    const float32_t2x2 rotationMatrix = float32_t2x2(cosAngleDifferenceHalf, -sinAngleDifferenceHalf, sinAngleDifferenceHalf, cosAngleDifferenceHalf);
```

Comment:
do these in a `else if (vertexIdx == 3u || vertexIdx == 0u)` to avoid duplication AND because the theads in a warp/wave go through the code in lockstep fashion and doing the same computation in different branches
https://jenni-westoby.github.io/Julia_GPU_examples/dev/Performance_thoughts/